### PR TITLE
Add OTLP exporter for OpenTelemetry

### LINKED ISSUE #47408: Distributed tracing via context propagation is broken after openTelemetry implementation
**What happened**:
Suppose there is a reverse proxy set up in front of self-hosted Grafana and the proxy injects a `uber-trace-id` header with a valid `uber-trace-id` value into proxies requests going to the Grafana, the Grafana tracing middleware failed to extract the trace id (tracing context) from the incoming request. This bug leads to that the context would create a brand new root span instead of keep using the trace id in the incoming request as the root trace id, in other words, the context propagation of distributed tracing is broken.

**What you expected to happen**:
Grafana tracing middleware should properly extract the trace id (trace context) from incoming requests and use it as a root trace id (trace context) so that users can search traces in the reverse proxy (or any other kind of services in front of Grafana) and in Grafana using the same propagated trace id for distributed tracing.

**How to reproduce it (as minimally and precisely as possible)**:
- enable tracing by either giving the address of [jaeger](https://grafana.com/docs/grafana/latest/administration/configuration/#address-1) or [opentelemetry jaeger](https://github.com/grafana/grafana/blob/38c842cf4ae13e9a83ded413095fc1fe28e26141/conf/defaults.ini#L972)
- send HTTP requests to Grafana e.g. using `curl` with a header `uber-trace-id` and a [valid `uber-trace-id` value](https://www.jaegertracing.io/docs/1.32/client-libraries/#value)
- for any context logs printed by Grafana, there would not be any `traceID` that matches the trace id in the `uber-trace-id` value of the incoming request

**Anything else we need to know?**:

**Environment**:
- Grafana version: versions after [8.4.0](https://github.com/grafana/grafana/releases/tag/v8.4.0) including this #42674
- Data source type & version:
- OS Grafana is installed on: WSL
- User OS & Browser: Chrome
- Grafana plugins:
- Others:

### PR DESCRIPTION
**Pull request by idafurjes**
**What this PR does / why we need it**:
On this PR we are adding configuration for tracing using OpenTelemetry with OTLP exporter and GRPC protocol. We are also adding configuration for text map propagation for OpenTelemetry for both jaeger and w3c format.

**Which issue(s) this PR fixes**:

Fixes #47408

**Special notes for your reviewer**:

### CODE REVIEW

**Review by zserge — APPROVED** [review]

**Review by ying-jeanne — APPROVED** [review]

File: pkg/infra/tracing/opentelemetry_tracing.go [hunk 1]
```diff
@@ -86,25 +118,82 @@ func (ots *Opentelemetry) initTracerProvider() (*tracesdk.TracerProvider, error)
...
+		resource.WithTelemetrySDK(),
+	)
+	if err != nil {
+		return nil, err
 	}
+
+	tp := tracesdk.NewTracerProvider(
+		tracesdk.WithBatcher(exp),
+		tracesdk.WithSampler(tracesdk.ParentBased(
+			tracesdk.AlwaysSample(),
```
Comment (jpkrohling) [inline pkg/infra/tracing/opentelemetry_tracing.go:144]: I'm not too familiar with how Grafana uses tracing but is there a switch somewhere to turn tracing off?

File: pkg/infra/tracing/test_helper.go [hunk 2]
```diff
@@ -1,7 +1,9 @@
 package tracing
 
 func InitializeTracerForTest() (Tracer, error) {
-	ots := &Opentelemetry{}
+	ots := &Opentelemetry{
+		enabled: "jaeger",
```
Comment (jpkrohling) [inline pkg/infra/tracing/test_helper.go:5]: I think there's one tracer for testing, keeping traces in memory and allowing for later retrieval. If this is interesting to you, let me know and I'll try to find more details.

**Review by jpkrohling — APPROVED** [review]
LGTM, just a couple of minor comments.

File: pkg/infra/tracing/opentelemetry_tracing.go [hunk 3]
```diff
@@ -6,18 +6,30 @@ import (
 	"time"
 
 	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/infra/log/level"
 	"github.com/grafana/grafana/pkg/setting"
+	"go.etcd.io/etcd/version"
```
Comment (bboreham) [inline pkg/infra/tracing/opentelemetry_tracing.go:11]: This is a typo, right?

Comment (idafurjes) [reply pkg/infra/tracing/opentelemetry_tracing.go:11]: yes, I missed it somehow

**Pull request merged**